palette/build/lut.rs

```diff
@@ -0,0 +1,651 @@
+            if i % 8 == 7 {
+                table.extend_from_slice("\n".as_bytes());
+            }
+        }
+
+        let table_name = format!("TO_{fn_type_uppercase}_U8");
+        writeln!(writer, "const {table_name}: [u32; {table_size}] = [").unwrap();
+        writer.write_all(&table).unwrap();
+        writeln!(writer, "\n];").unwrap();
+
+        let entry_shift = 23 - man_index_size;
+        let man_shift = 15 - man_index_size;
+
+        writeln!(
+            writer,
+            "impl FromLinear<f64, u8> for {fn_type} {{\
+            \n\t#[inline]\
+            \n\tfn from_linear(linear: f64) -> u8 {{\
+            \n\t\t{fn_type}::from_linear(linear as f32)\
+            \n\t}}\
+            \n}}"
+        )
+        .unwrap();
+
+        let min_float_string = format!("{:#x}", min_float_bits);
+        writeln!(
+            writer,
+            "impl FromLinear<f32, u8> for {fn_type} {{\
+                \n\t#[inline]\
+                \n\tfn from_linear(linear: f32) -> u8 {{\
+                    \n\t\tconst MAX_FLOAT_BITS: u32 = 0x3f7fffff; // 1.0 - f32::EPSILON\
+                    \n\t\tconst MIN_FLOAT_BITS: u32 = {min_float_string}; // 2^(-{exp_table_size})\
+                    \n\t\tlet max_float = f32::from_bits(MAX_FLOAT_BITS);\
+                        \n\t\tlet min_float = f32::from_bits(MIN_FLOAT_BITS);\
+                        \n\n\t\tlet mut input = linear;\
+                    \n\t\tif input.partial_cmp(&min_float) != Some(core::cmp::Ordering::Greater) {{\
+                        \n\t\t\tinput = min_float;\
+                    \n\t\t}} else if input > max_float {{\
+                        \n\t\t\tinput = max_float;\
+                    \n\t\t}}
+                    \n\t\tlet input_bits = input.to_bits();\
+                    \n\t\t#[cfg(test)]\
+                    \n\t\t{{\
+                        \n\t\t\tdebug_assert!((MIN_FLOAT_BITS..=MAX_FLOAT_BITS).contains(&input_bits));\
+                    \n\t\t}}\
+                    \n\n\t\tlet entry = {{\
+                        \n\t\t\tlet i = ((input_bits - MIN_FLOAT_BITS) >> {entry_shift}) as usize;\
+                        \n\t\t\t#[cfg(test)]\
+                        \n\t\t\t{{\
+                            \n\t\t\t\tdebug_assert!({table_name}.get(i).is_some());\
+                        \n\t\t\t}}\
+                        \n\t\t\tunsafe {{ *{table_name}.get_unchecked(i) }}\
+                    \n\t\t}};\
+                    \n\t\tlet bias = (entry >> 16) << 9;\
+                    \n\t\tlet scale = entry & 0xffff;\
+                    \n\n\t\tlet t = (input_bits >> {man_shift}) & 0xff;\
+                    \n\t\tlet res = (bias + scale * t) >> 16;\
+                    \n\t\t#[cfg(test)]\
+                    \n\t\t{{\
+                        \n\t\t\tdebug_assert!(res < 256, \"{{}}\", res);\
+                    \n\t\t}}\
+                    \n\t\tres as u8\
+                \n\t}}\
+            \n}}"
```

Comment:
This is the part I think we don't need to generate every time. It looks to me like it could be a library function that takes a table reference and input value.